### PR TITLE
feature: add helper method to generate pipeline adjacency list

### PR DESCRIPTION
**Pull request by nmadan**
*Issue #, if available:*

*Description of changes:* Added a helper method to create adjacency list representing the step dependency DAG within a sagemaker pipeline

*Testing done:* unit tests

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backword compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [ ] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [ ] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [ ] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker/workflow/properties.py [hunk 1]
```diff
@@ -98,17 +110,21 @@ def expr(self):
...
             path (str): The parent path of the PropertiesList instance.
             shape_name (str): The botocore service model shape name.
             service_name (str): The botocore service name.
         """
-        super(PropertiesList, self).__init__(path, shape_name)
+        super(PropertiesList, self).__init__(step_name, path, shape_name)
+        self.step_name = step_name
         self.shape_name = shape_name
         self.service_name = service_name
+        self.path = path
```
Comment (qidewenwhen) [inline src/sagemaker/workflow/properties.py:127]: So the `self.path` is a relative path without the root: `f"Steps.{step_name}."` while the `self._path` (from super class) is an absolute path with the root right?
If so can we add some comments or update the attribute name to make it more clear?

File: src/sagemaker/workflow/steps.py [hunk 2]
```diff
@@ -148,6 +148,33 @@ def _resolve_depends_on(
...
+                    dependencies.update(_find_dependencies_in_step_arguments(item))
+            return dependencies
+
+        step_dependencies = set()
+        if self.depends_on:
+            for depends_on_step in self.depends_on:
+                if isinstance(depends_on_step, str):
+                    step_dependencies.add(depends_on_step)
+                elif isinstance(depends_on_step, Step):
+                    step_dependencies.add(depends_on_step.name)
```
Comment (qidewenwhen) [inline src/sagemaker/workflow/steps.py:174]: What if a StepCollection is in the depends on list?
This [merged commit](https://github.com/aws/sagemaker-python-sdk/commit/79ecab28a5004057c0842cfa291ef964d1f3129b#diff-f07f30c98bafb73739b39e93626762453e4249ce4f0b19c95692369397131378R141-R143) just supported the StepCollection to be depended on.
Can you add some tests on this as well?

File: src/sagemaker/workflow/step_collections.py [hunk 3]
```diff
@@ -54,6 +54,13 @@ def properties(self):
         size = len(self.steps)
         return self.steps[size - 1].properties
 
+    def _find_step_dependencies(self):
+        """Find the step names this step collection is dependent on."""
+        dependencies = {}
+        for step in self.steps:
+            dependencies[step.name] = step._find_step_dependencies()
+        return dependencies
+
```
Comment (qidewenwhen) [inline src/sagemaker/workflow/step_collections.py:63]: This [merged commit](https://github.com/aws/sagemaker-python-sdk/commit/79ecab28a5004057c0842cfa291ef964d1f3129b#diff-043e8c79b0b9ccd41f8164a29ccc1f9e67f38e349aafffb96cd91762c3824bc7R49-R56) just supports StepCollection properties.
What if another step is referencing the StepCollection properties? Would the last sub step name be appended in the adjacency list?
Please point me to a test of this case if any

File: src/sagemaker/workflow/steps.py [hunk 4]
```diff
@@ -148,6 +148,33 @@ def _resolve_depends_on(
                 raise ValueError(f"Invalid input step name: {step}")
         return depends_on
 
+    def _find_step_dependencies(self):
+        """Find the step names this step is dependent on."""
+
+        def _find_dependencies_in_step_arguments(obj):
```
Comment (navaj0) [inline src/sagemaker/workflow/steps.py:154]: Better to move this out. You'll need it when run the pipeline

```
def _find_property_references_in_step_arguments() -> Set[Properties]:
    ...
```

File: src/sagemaker/workflow/steps.py (hunk 2 above)
Comment (navaj0) [reply src/sagemaker/workflow/steps.py:174]: Right. The StepCollection needs to be treated with care.

File: src/sagemaker/workflow/pipeline.py [hunk 5]
```diff
@@ -322,6 +322,16 @@ def _interpolate_step_collection_name_in_depends_on(self, step_requests: dict):
...
 
+    def _generate_adjacency_list(self):
+        """Generate an adjacency list representing the step dependency DAG in this pipeline."""
+        adjacency_list = {}
+        for step in self.steps:
+            if isinstance(step, Step):
+                adjacency_list[step.name] = step._find_step_dependencies()
+            elif isinstance(step, StepCollection):
+                adjacency_list.update(step._find_step_dependencies())
+        return adjacency_list
```
Comment (navaj0) [inline src/sagemaker/workflow/pipeline.py:333]: Suggest we create a new class and offload the building of the graph, validation (no cycle), traversal to this class

```
class _PipelineGraph:
    def __init__ (self, steps: List[Step]): 
        self.adjacency_list = {}
        ...
    def __iter__():
        ...
    def __next()__() -> List[Step]:
        ...

```

Comment (navaj0) [reply src/sagemaker/workflow/pipeline.py:333]: Condition step needs to be treated in a special way. See
https://code.amazon.com/packages/IronmanTiogaPipelineDefinitionRepository/blobs/8b985bed9418f5cace3f7bea553e4354752c61f3/--/src/com/amazon/ironmantiogapipelinerepository/definition/PipelineDefinition.java#L73

File: src/sagemaker/workflow/pipeline.py [hunk 6]
```diff
@@ -534,3 +535,138 @@ def wait(self, delay=30, max_attempts=60):
...
+
+    Attributes:
+        steps (Sequence[Union[Step, StepCollection]]): Sequence of `Step`s and/or `StepCollection`s
+            that represent each node in the pipeline DAG
+    """
+
+    def __init__(self, steps: Sequence[Union[Step, StepCollection]]):
+        if self.is_duplicate_step_name(steps):
+            raise ValueError("Pipeline steps cannot have duplicate names.")
+        self.step_map = PipelineGraph._generate_step_map(steps)
```
Comment (navaj0) [inline src/sagemaker/workflow/pipeline.py:551]: These two could be merged into one.

File: src/sagemaker/workflow/pipeline.py [hunk 7]
```diff
@@ -534,3 +535,138 @@ def wait(self, delay=30, max_attempts=60):
...
+            if isinstance(step, Step):
+                step_map[step.name] = step
+                if isinstance(step, ConditionStep):
+                    step_map.update(
+                        PipelineGraph._generate_step_map(step.if_steps + step.else_steps)
+                    )
+            elif isinstance(step, StepCollection):
+                step_map[step.name] = step
+                for inner_step in step.steps:
+                    step_map[inner_step.name] = inner_step
```
Comment (navaj0) [inline src/sagemaker/workflow/pipeline.py:572]: A StepCollection is literally a list of steps. It is just an abstraction to the users. When building the graph, we just need to take care of the steps inside.

File: src/sagemaker/workflow/pipeline.py [hunk 8]
```diff
@@ -534,3 +535,138 @@ def wait(self, delay=30, max_attempts=60):
...
+        from collections import defaultdict
+
+        dependency_list = defaultdict(set)
+        for step in self.step_map.values():
+            if isinstance(step, Step):
+                dependency_list[step.name].update(step._find_step_dependencies(self.step_map))
+            elif isinstance(step, StepCollection):
+                step_collection_dependencies = step._find_step_dependencies(self.step_map)
+                for k, v in step_collection_dependencies.items():
+                    dependency_list[k].update(v)
```
Comment (navaj0) [inline src/sagemaker/workflow/pipeline.py:591]: We can skip the parent StepCollection here.

File: src/sagemaker/workflow/pipeline.py [hunk 9]
```diff
@@ -534,3 +535,138 @@ def wait(self, delay=30, max_attempts=60):
...
+
+            if isinstance(step, ConditionStep):
```
Comment (navaj0) [inline src/sagemaker/workflow/pipeline.py:593]: The if else branches in the condition step are already flattened out. So it just need to be treated as a single node. 

The condition step itself has condition expressions, which may contain property references.

File: src/sagemaker/workflow/pipeline.py [hunk 10]
```diff
@@ -534,3 +535,126 @@ def wait(self, delay=30, max_attempts=60):
...
+        step_map = {}
+        for step in steps:
+            if isinstance(step, Step):
+                if step.name in step_map:
+                    raise ValueError("Pipeline steps cannot have duplicate names.")
+                step_map[step.name] = step
+                if isinstance(step, ConditionStep):
+                    step_map.update(
+                        PipelineGraph._generate_step_map(step.if_steps + step.else_steps)
+                    )
```
Comment (navaj0) [inline src/sagemaker/workflow/pipeline.py:568]: The names of steps in if/else branches can't conflict with steps outside. So it is better to do something like 

```
for step in steps:
    if step.name in self.step_map:
        raise ValueError
    self.step_map[step.name] = step
    if isinstance(step, ConditionStep):
        self.generate_step_map(...)
    elif isinstance(step, StepCollection):
        self.generate_step_map(step.steps) 
```

File: src/sagemaker/workflow/pipeline.py (hunk 9 above)
Comment (navaj0) [reply src/sagemaker/workflow/pipeline.py:593]: This comment can be ignored.

File: tests/unit/sagemaker/workflow/test_pipeline_graph.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    session_mock.default_bucket = Mock(name="default_bucket", return_value="s3_bucket")
+    return session_mock
+
+
+@pytest.fixture
+def role_arn():
+    return "arn:role"
+
+
+def test_pipeline_duplicate_step_name(sagemaker_session_mock):
```
Comment (navaj0) [inline tests/unit/sagemaker/workflow/test_pipeline_graph.py:35]: A couple of more scenarios:
* duplicate_step_name_in_condition_step
* duplicate_step_name_in_step_collection

File: tests/unit/sagemaker/workflow/test_pipeline_graph.py [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+    step2 = CustomStep(name="foo")
+    pipeline = Pipeline(
+        name="MyPipeline", steps=[step1, step2], sagemaker_session=sagemaker_session_mock
+    )
+    with pytest.raises(ValueError) as error:
+        PipelineGraph.from_pipeline(pipeline)
+    assert "Pipeline steps cannot have duplicate names." in str(error.value)
+
+
+def test_pipeline_graph_acyclic(sagemaker_session_mock):
```
Comment (navaj0) [inline tests/unit/sagemaker/workflow/test_pipeline_graph.py:46]: more scenarios with condition steps
* test_pipeline_graph_acyclic_with_condition_step
  * CustomStep -> ConditionStep/IfSteps/ElseStep: dependencies are defined via property reference
  * CustomStep -> ConditionStep/IfStep/ElseStep: dependencies are defined via dependsOn attribute
* test_pipeline_graph_acyclic_with_step_collection
  * CustomStep -> ModelStep -> CustomStep: dependencies are defined via property reference
  * CustomStep -> ModelStep -> CustomStep: dependencies are defined via dependsOn attribute

File: src/sagemaker/workflow/entities.py [hunk 13]
```diff
@@ -102,3 +102,8 @@ def to_string(self):
     @abc.abstractmethod
     def expr(self) -> RequestType:
         """Get the expression structure for workflow service calls."""
+
+    @property
+    @abc.abstractmethod
+    def depends_on(self) -> List[str]:
```
Comment (navaj0) [inline src/sagemaker/workflow/entities.py:108]: Can we make this a protected method
```
_referenced_steps
```

**Review by navaj0 — APPROVED** [review]

File: tests/unit/sagemaker/workflow/test_condition_step.py [hunk 14]
```diff
@@ -60,3 +77,26 @@ def test_condition_step():
...
+    pipeline = Pipeline(
+        name="MyPipeline",
+        steps=[step_cond],
+        sagemaker_session=sagemaker_session,
+        parameters=[param],
+    )
+    adjacency_list = PipelineGraph.from_pipeline(pipeline).adjacency_list
+    assert ordered(adjacency_list) == ordered(
+        {"CondStep": ["IfStep", "ElseStep"], "IfStep": [], "ElseStep": []}
+    )
```
Comment (qidewenwhen) [inline tests/unit/sagemaker/workflow/test_condition_step.py:102]: From visualization perspective, seems it's not able to tell which steps, in adjacency list of "CondStep", are in IF branch and which are in ELSE branch.
Please correct me if I'm wrong

Comment (qidewenwhen) [reply tests/unit/sagemaker/workflow/test_condition_step.py:102]: Nic's design doc uses the following adjacency list format:
```
[
        {"StepName": 'AbaloneProcess',
         "OutBoundEdges": [
            {
                "nextStepName": "AbaloneTrain",
                "edgeLabel": None                <<<<<<<<<<< indicating the IF or ELSE branch
            },
            {
                "nextStepName": "AbaloneEval",
                "edgeLabel": None
            }
         ]
        },
        {"StepName": 'AbaloneMSECond',
         "OutBoundEdges": [
            {
                "nextStepId": "AbaloneRegisterModel",
                "edgeLabel": "true"
            },
            {
                "nextStepId": "AbaloneCreateModel",
                "edgeLabel": "true"
            },
            {
                "nextStepId": "AbaloneMSEFail",
                "edgeLabel": "false"
            }
         ]
        },
```

Can we also apply it here?

Comment (nmadan) [reply tests/unit/sagemaker/workflow/test_condition_step.py:102]: This format can be derived using a combination of the current adjacency list format and the `pipeline.steps` field.
Same thing applies to the condition step if-else branches. The goal of the adjacency list to give parent-child step relationship info. Nic can use the adjacency list + pipeline.steps to decipher whether the child of a condition step is a if-child or else-child.

**Review by qidewenwhen — APPROVED** [review]

File: src/sagemaker/workflow/pipeline.py [hunk 15]
```diff
@@ -534,3 +535,113 @@ def wait(self, delay=30, max_attempts=60):
...
+                for child_step in step.if_steps + step.else_steps:
+                    if isinstance(child_step, Step):
+                        dependency_list[child_step.name].add(step.name)
+                    elif isinstance(child_step, StepCollection):
+                        child_first_step = self.step_map[child_step.name].steps[0].name
+                        dependency_list[child_first_step].add(step.name)
+
+        adjacency_list = {}
+        for step in dependency_list:
+            for step_dependency in dependency_list[step]:
```
Comment (mufaddal-rohawala) [inline src/sagemaker/workflow/pipeline.py:590]: nit: prb can be one line `for step, step_dependency in dependency_list.items()`

File: src/sagemaker/workflow/steps.py [hunk 16]
```diff
@@ -148,9 +159,51 @@ def _resolve_depends_on(
...
+                        dependencies.update(map(_get_step_name_from_str, value._referenced_steps))
+                    dependencies.update(_find_dependencies_in_step_arguments(value))
+            elif isinstance(obj, list):
+                for item in obj:
+                    if isinstance(item, (PipelineVariable, Condition)):
+                        dependencies.update(map(_get_step_name_from_str, item._referenced_steps))
+                    dependencies.update(_find_dependencies_in_step_arguments(item))
+            return dependencies
+
+        def _find_dependencies_in_depends_on_list():
```
Comment (mufaddal-rohawala) [inline src/sagemaker/workflow/steps.py:184]: nit: This func has grown too long , can possibly move out all these internal functions? Should also help in Unit testing?

File: src/sagemaker/workflow/pipeline.py (hunk 15 above)
Comment (nmadan) [reply src/sagemaker/workflow/pipeline.py:590]: dependency_list is a Dict[str,List[str]]. This line is looping over the each item in the List.

**Review by shreyapandit — APPROVED** [review]

**Review by mufaddal-rohawala — APPROVED** [review]

**Pull request merged**